---
title: "BIMM 143 Lab 15 Mi Tran (2)"
author: "Mi Tran A15556214"
date: "11/23/2021"
output: pdf_document
---
The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

# 1. Bioconductor and DESeq2 setup
```{r}
# Run these codes one at a time to install the core Bioconductor packages
# install.packages("BiocManager")
# BiocManager::install()
# BiocManager::install("DESeq2")
```

```{r}
# library(BiocManager)
# library(DESeq2)
```

# 2. Import countData and colData

First, download the input 'airway_scaledcounts.csv' and 'airway_metadata.csv' 
```{r}
counts <- read.csv("https://bioboot.github.io/bimm143_F21/class-material/airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("https://bioboot.github.io/bimm143_F21/class-material/airway_metadata.csv")
```

Now, take a look at each.
```{r}
head(counts)
head(metadata)
```

Sanity check on correspondence of 'counts' and 'metadata'. 
```{r}
all ( metadata$id == colnames(counts) )
```

> Q1. How many genes are in this data set? **38694 genes**

There are 'r nrow(counts)' genes in this dataset. 
```{r}
nrow(counts)
```

> Q2. How many 'control' cell lines do we have? **4 'control' cell lines**

```{r}
n.control <- sum (metadata$dex == "control" )
n.control
```

# 3. Toy Differential Gene Expression

Let's perform some exploratory differential gene expression analysis. **Note: This analysis is for demonstration only. NEVER do differential expression analysis this way!**

Note that the control samples are SRR1039508, SRR1039512, SRR1039516, SRR1039520. This bit of code will first find the sample 'id' for those labelled control. Then calculate the mean counts per gene across these samples: 

```{r}
control <- metadata[metadata$dex == "control", ]
control.counts <- counts[ , control$id]
control.mean <- rowSums( control.counts )/4
head(control.mean)
```

**Side-note:** An alternative way to do this same thing using the dplyr package from the tidyverse is shown below. Which do you prefer and why? 

```{r}
library(dplyr)
control <- metadata %>% filter(dex=="control")
control.counts <- counts %>% select(control$id)
control.mean <- rowSums(control.counts)/4
head(control.mean)
```

> Q3. How would you make the above code in either approach more robust? **In the above approaches, 'rowSums()/4' was used to find the average value. We should change it into rowMeans().**

```{r}
control <- metadata[metadata$dex == "control",]
control.counts <- counts[ , control$id]
control.mean <-rowMeans(control.counts)
head(control.mean)
```

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called 'treated.mean')

```{r}
treated <- metadata[metadata[,"dex"] == "treated",]
treated.counts <- counts[ ,treated$id]
treated.mean <- rowSums( treated.counts )/4
head(treated.mean)
```

We will combine our meancount data for bookkeeping purposes. 

```{r}
meancounts <- data.frame(control.mean, treated.mean)
#meancounts
```

> Q5. a. Create a scatter plot showing the mean of the treated samples against the mean of the control samples. 

```{r}
plot(meancounts[,1], meancounts[,2], xlab="Control", ylab="Treated")
```

> Q5. b. You could also use the ggplot2 package to make this figure producing the plot below. What geom function would you use for this plot? **Using geom_point()**

```{r}
library(ggplot2)
ggplot(meancounts) +
  aes(control.mean, treated.mean) +
  geom_point()
```

> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this? 

```{r}
plot(meancounts[,1], meancounts[,2], log = "xy", 
     xlab= "log countrol counts",
     ylab= "log of treated counts" )
```

So let's add a log2 fold change column to our results.
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

There are a couple of "weird" results. Namely, the NaN ("not a number") and --Inf (negative infinity) results. 

The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. 

```{r}
zero.vals <- which(meancounts[,1:2] == 0, arr.ind = TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

> Q7. What is the purpose of the arr.ind argument in which the which() function call above? Why would we then take the first column of the output and need to call the unique() function? 
**The arr.ind argument causes which() to return both the row and column indices (ie. positions) where there are TRUE values, in which here it tells us which genes (rows) and samples (columns) have zero counts. We're ignoring any genes that have zero counts in any sample so we just focus on the row answer. We need to call the unique() function to ensure that we don't count any row twice if it has zero entries in both samples.**

The common threshold used for calling something differential expressed is a log2(FoldChange) of greater than 2 or less than -2. Let's filter the data set both ways to see how many genes are up or down regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level? **250 up regulated genes**

```{r}
sum(up.ind == "TRUE")
```

> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level? **367 down regulated genes**

```{r}
sum(down.ind == "TRUE")
```

> Q10. Do you trust these results? Why or why not?
**No, these results are misleading because all of our analysis has been done based on fold change, and fold change can be large without being statistically significant. We have yet to determine whether the differences we are seeing are significant.**

# 4. DESeq2 Analysis

```{r}
library(DESeq2)
citation("DESeq2")
```

## Importing Data

Take a look at metadata again. The thing we’re interested in is the dex column, which tells us which samples are treated with dexamethasone versus which samples are untreated controls. We’ll specify the design with a tilde, like this: design = ~dex.

We will use the DESeqDataSetFromMatrix() function to build the required DESeqDataSet object and call it dds, short for our DESeqDataSet. If you get a warning about “some variables in design formula are characters, converting to factors” don’t worry about it. Take a look at the dds object once you create it.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata, 
                              design =~dex)
dds
```

## DESeq Analysis
Here, we’re running the DESeq pipeline on the dds object, and reassigning the whole thing back to dds, which will now be a DESeqDataSet populated with all those values. 

```{r}
dds <- DESeq(dds)
```

## Getting results

Since we’ve got a fairly simple design (single factor, two groups, treated versus control), we can get results out of the object simply by calling the results() function on the DESeqDataSet that has been run through the pipeline.

```{r}
res <- results(dds)
res
```

Convert the res object to a data.frame with the as.data.frame() function and then pass it to View() to bring it up in a data viewer.

We can summarize some basic tallies using the summary function.

```{r}
summary(res)
```

The results function contains a number of arguments to customize the results table. By default the argument alpha is set to 0.1. If the adjusted p value cutoff will be a value other than 0.1, alpha should be set to that value:

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

# 5. Adding Annotation Data

Our result table so far only contains the Ensembl gene IDs. However, alternative gene names and extra annotation are usually required for informative interpretation of our results. In this section we will add this necessary annotation data to our results.

We will use one of Bioconductor’s main annotation packages to help with mapping between various ID schemes. Here we load the AnnotationDbi package and the annotation data package for humans org.Hs.eg.db.

```{r}
#BiocManager::install("AnnotationDbi")
#BiocManager::install("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")
```

The later of these is is the organism annotation package (“org”) for Homo sapiens (“Hs”), organized as an AnnotationDbi database package (“db”), using Entrez Gene IDs (“eg”) as primary key. To get a list of all available key types that we can use to map between, use the columns() function:
```{r}
columns(org.Hs.eg.db)
```

The main function we will use from the AnnotationDbi package is called mapIds().

We can use the mapIds() function to add individual columns to our results table. We provide the row names of our results table as a key, and specify that keytype=ENSEMBL. The column argument tells the mapIds() function which information we want, and the multiVals argument tells the function what to do if there are multiple possible values for a single input value. Here we ask to just give us back the first one that occurs in the database.

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(res)
```

> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

head(res)
```

You can arrange and view the results by the adjusted p-value
```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])
```

Finally, let’s write out the ordered significant results with annotations. 

```{r}
write.csv(res[ord,], "deseq_results.csv")
```

# 6. Data Visualization

## Volcano Plot
```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```

To make this more useful we can add some guidelines (with the abline() function) and color (with a custom color vector) highlighting genes that have padj<0.05 and the absolute log2FoldChange>2.

```{r}
plot( res$log2FoldChange,  -log(res$padj), 
 ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```

To color the points we will setup a custom color vector indicating transcripts with large fold change and significant differences between conditions:

```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

For even more customization you might find the EnhancedVolcano bioconductor package useful (Note. It uses ggplot under the hood):

First we will add the more understandable gene symbol names to our full results object res as we will use this to label the most interesting genes in our final plot.

```{r}
#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
x <- as.data.frame(res)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```

# 7. Pathway Analysis

## Pathway Analysis with R and Bioconductor
```{r}
# BiocManager::install( c("pathview", "gage", "gageData"))
library(pathview)
library(gage)
library(gageData)
data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
        
```
The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now let's run the gage pathway analysis.

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

The top three Kegg pathways indicated here include Graft-versus-host disease, Type I diabetes and the Asthma pathway (with pathway ID hsa05310).

Now, let’s try out the pathview() function from the pathview package to make a pathway plot with our RNA-Seq expression results shown in color.
To begin with lets manually supply a pathway.id (namely the first part of the "hsa05310 Asthma") that we could see from the print out above.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```
```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa05310", kegg.native=FALSE)
```

